catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,465 @@
+        SlotDuration as Block0SlotDuration,
+    },
+    time::SecondsSinceUnixEpoch,
+};
+use jormungandr_testing_utils::wallet::Wallet;
+use std::collections::{HashMap, HashSet};
+
+#[allow(clippy::large_enum_variant)]
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    DeserializeError(#[from] jormungandr_lib::interfaces::FragmentLogDeserializeError),
+
+    #[error(transparent)]
+    LedgerError(#[from] chain_impl_mockchain::ledger::Error),
+
+    #[error("Couldn't initiate a new wallet")]
+    WalletError(#[from] jormungandr_testing_utils::wallet::WalletError),
+
+    #[error(transparent)]
+    Block0ConfigurationError(#[from] jormungandr_lib::interfaces::Block0ConfigurationError),
+
+    #[error("Block0 do not contain any voteplan")]
+    MissingVoteplanError,
+
+    #[error("Could not verify transaction {id} signature with range {range:?}")]
+    InvalidTransactionSignature {
+        id: String,
+        range: std::ops::Range<i32>,
+    },
+}
+
+fn timestamp_to_system_time(ts: SecondsSinceUnixEpoch) -> SystemTime {
+    SystemTime::UNIX_EPOCH.add(Duration::new(ts.to_secs(), 0))
+}
+
+fn fragment_log_timestamp_to_blockdate(
+    timestamp: SecondsSinceUnixEpoch,
+    timeframe: &TimeFrame,
+    ledger: &Ledger,
+) -> Option<BlockDate> {
+    let slot = timestamp_to_system_time(timestamp);
+    let new_slot = timeframe.slot_at(&slot)?;
+    let epoch_position = ledger.era().from_slot_to_era(new_slot)?;
+    Some(BlockDate::from(epoch_position))
+}
+
+fn timeframe_from_block0_start_and_slot_duration(
+    block0_start: SecondsSinceUnixEpoch,
+    slot_duration: Block0SlotDuration,
+) -> TimeFrame {
+    let timeline = Timeline::new(timestamp_to_system_time(block0_start));
+
+    TimeFrame::new(
+        timeline,
+        SlotDuration::from_secs(<u8>::from(slot_duration) as u32),
+    )
+}
+
+fn committee_id_to_address(id: CommitteeIdDef) -> Address {
+    let id = CommitteeId::from(id);
+    let pk = id.public_key();
+    chain_addr::Address(Discrimination::Production, Kind::Account(pk)).into()
+}
+
+fn voteplans_from_block0(block0: &Block) -> HashMap<VotePlanId, VotePlan> {
+    block0
+        .fragments()
+        .filter_map(|fragment| {
+            if let Fragment::VotePlan(tx) = fragment {
+                let voteplan = tx.as_slice().payload().into_payload();
+                Some((voteplan.to_id(), voteplan))
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+fn verify_original_tx(
+    spending_counter: SpendingCounter,
+    block0_hash: &HeaderId,
+    sign_data_hash: &TransactionSignDataHash,
+    account: &account::Identifier,
+    witness: &account::Witness,
+    range_check: Range<i32>,
+) -> (bool, u32) {
+    let spending_counter: i32 = <u32>::from(spending_counter) as i32;
+    for i in range_check {
+        let new_spending_counter = spending_counter.add(i).clamp(0, i32::MAX) as u32;
```

Comment:
It's probably slower if you start with `spending_counter - 50` if the actual drift tends to be around the expected value. I'd do iteration over `n`, `n + 1`, `n - 1`, ...

Also, using `clamp` means you might do redundant checks. Don't cast to a `i32`, but use something like:

```rust
    if let Some(n) = spending_counter.checked_add(i) {
        // try with n as spending counter
    } else {
        continue;
    }
    if let Some(n) = spending_counter.checked_sub(i) {
        // try with n as spending counter
    } else {
        continue;
    }
```